central/resourcecollection/datastore/datastore_impl.go

```diff
@@ -119,6 +132,33 @@ func (ds *datastoreImpl) deleteCollectionFromGraph(id string) error {
 	return ds.graph.DeleteVertex(id)
 }
 
+func (ds *datastoreImpl) dryRunCollectionInGraph(obj *storage.ResourceCollection) error {
+	ds.graphLock.Lock()
+	defer ds.graphLock.Unlock()
+
+	var ret error
+
+	// we essentially just add the obj to the graph and delete it before returning
+	err := ds.graph.AddVertexByID(obj.GetId(), obj.GetName())
+	if err != nil {
+		return err
+	}
+	for _, edge := range obj.GetEmbeddedCollections() {
+		err = ds.graph.AddEdge(obj.GetId(), edge.GetId())
+		if err != nil {
+			ret = err
+			break
+		}
+	}
+
+	err = ds.graph.DeleteVertex(obj.GetId())
+	if err != nil && ret == nil {
+		ret = err
+	}
```

Comment:
This can potentially result in a lock timeout. The chances of that happening increase significantly as the graph size increases.
Have you experimented with the `Copy()` operation performance? That is create a copy of DAG and work on the copy and set it to nil.

Reply:
These are good ideas and I've captured them for future work: https://issues.redhat.com/browse/ROX-13228

I think it would be good to continue working through functionality and getting up the service layer and getting baseline workflow functionality down before spending too much time on optimizing

Reply:
Fair enough